planner_cspace/src/planner_3d.cpp

```diff
@@ -1275,29 +1275,77 @@ class Planner3dNode
     act_->start();
     act_tolerant_->start();
   }
-  void spin()
+
+  GridAstarModel3D::Vec pathPose2Grid(const geometry_msgs::PoseStamped& pose) const
   {
-    ros::Rate wait(freq_);
-    ROS_DEBUG("Initialized");
+    GridAstarModel3D::Vec grid_vec;
+    grid_metric_converter::metric2Grid(map_info_, grid_vec[0], grid_vec[1], grid_vec[2],
+                                       pose.pose.position.x, pose.pose.position.y, tf2::getYaw(pose.pose.orientation));
+    grid_vec.cycleUnsigned(map_info_.angle);
+    return grid_vec;
+  }
 
+  void waitUntil(const ros::Time& next_replan_time, const nav_msgs::Path& previous_path)
+  {
     while (ros::ok())
     {
-      wait.sleep();
+      const ros::Time prev_map_update_stamp = last_costmap_;
       ros::spinOnce();
-
-      const ros::Time now = ros::Time::now();
+      const bool costmap_udpated = last_costmap_ != prev_map_update_stamp;
 
       if (has_map_)
       {
         updateStart();
 
         if (jump_.detectJump())
+        {
           bbf_costmap_.clear();
+          // Robot pose jumped.
+          return;
+        }
 
-        if (!goal_updated_ && has_goal_)
-          updateGoal();
+        if (costmap_udpated && previous_path.poses.size() > 1)
+        {
+          const GridAstarModel3D::VecWithCost start_grid(pathPose2Grid(previous_path.poses.front()));
+          const GridAstarModel3D::Vec last_grid = pathPose2Grid(previous_path.poses.back());
+          GridAstarModel3D::Vec prev_grid = start_grid.v_;
+          for (size_t i = 1; i < previous_path.poses.size(); ++i)
+          {
+            const GridAstarModel3D::Vec cur_grid = pathPose2Grid(previous_path.poses[i]);
+            if (model_->cost(prev_grid, cur_grid, {start_grid}, last_grid) < 0)  // NOLINT(whitespace/braces)
```

Comment:
Do we need previous grid?
`previous_path` is already interpolated with the resolution higher than grid.
`cm_[cur_grid] == 100` would be enough.

Reply:
Done